### PR TITLE
Sol to UML initial

### PR DESCRIPTION
**Pull request by DanglSan**

### CODE REVIEW

**Comment by DanglSan** [comment]
Помимо прочего надо бы написать какие-то тесты, но я, если честно, слабо представляю, какие, поэтому прошу совета.

File: .gitignore [hunk 1]
```diff
@@ -1 +1,21 @@
...
+.vscode
+.project
+.favorites.json
+.settings/
+
+## Build
 /target
+
+.pre-commit-config.yaml
+config.toml
```
Comment (leviska) [inline .gitignore]: ?

File: .gitignore [hunk 2]
```diff
@@ -1 +1,21 @@
+## File system
+.DS_Store
+desktop.ini
+
+## Editor
```
Comment (leviska) [inline .gitignore:5]: fyi there is a gitignore, which is usually used for ignoring project related files, which are shared with everyone
And there is private ignore rules, https://stackoverflow.com/questions/6117109/is-it-possible-to-have-a-custom-gitignore-read-only-access, which are private for each user, and not shared, and my recommendation is to put your ide specific folders into it (because it's hard and stupid to try to support every ide files in gitignore)

So, basically rust inits gitignore with only `target/`, and it should be enough, so you can try to make gitignore as such, and configure exclude to fit you, although it isn't very important and you could leave this file as is

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,39 @@
+FROM rust:1.59 as build
+
+# create a new empty shell project
+RUN USER=root cargo new --bin sol_to_uml
```
Comment (leviska) [inline Dockerfile]: why do you need `USER=root`?

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,39 @@
+FROM rust:1.59 as build
```
Comment (leviska) [inline Dockerfile]: `rust:1-whatever version of ubuntu you're using afterwards`
(For example just `rust:1-ubuntu`)
We dont really care about strict rust version, so we're allow docker to update to newer rust versions
And it's best to specify exact ubuntu version, so rust will build and link to the proper architecture and libraries

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+# create a new empty shell project
+RUN USER=root cargo new --bin sol_to_uml
+WORKDIR /sol_to_uml
+
+# copy over your manifests
+COPY ./Cargo.lock ./Cargo.lock
+COPY ./Cargo.toml ./Cargo.toml
+#COPY Cargo.toml ./sol_to_uml/Cargo.toml
```
Comment (leviska) [inline Dockerfile]: remove

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+# this build step will cache your dependencies
+RUN cargo build --release
```
Comment (leviska) [inline Dockerfile]: (even if you copied this from the net) nicely done

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+RUN rm ./src/*.rs
```
Comment (leviska) [inline Dockerfile]: Do we need it? If so, may be then `rm -rf ./src`

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+# copy your source tree
+COPY ./src ./src
+
+# build for release
+RUN rm ./target/release/deps/sol_to_uml*
```
Comment (leviska) [inline Dockerfile]: Do we need this?

File: Dockerfile (hunk 3 above)
Comment (leviska) [reply Dockerfile]: why not just mkdir? Or even workdir?

File: .gitignore (hunk 2 above)
Comment (DanglSan) [reply .gitignore:5]: I just copied .gitignore from blockscout main repository, that`s the reason it looks like this. So, I think, we can leave it as is for uniformity.

File: Dockerfile (hunk 4 above)
Comment (DanglSan) [reply Dockerfile]: Didn`t find rust ubuntu docker image. As I understand, official rust image is based on Debian, so I set bullseye version. Also set architecture amd64 (but not sure if it needed) and made image slim.

File: Dockerfile (hunk 3 above)
Comment (DanglSan) [reply Dockerfile]: USER=root comes from the guide I used to create this Dockerfile. As I can see it's not required, so I removed it. 'cargo new' is mainly used to create main.rs file in src cause I don't copy ./src in the first step but should run 'cargo build'.

File: Dockerfile (hunk 8 above)
Comment (DanglSan) [reply Dockerfile]: Probably not, also comes from the guide. Docker works fine without it.

File: Dockerfile [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+FROM node:16-bullseye-slim
+
+RUN apt-get update
+
+WORKDIR /usr/src/
+
+# sol2uml needed phantom which installation needed bzip2
+RUN apt install bzip2
+RUN npm install phantom
+RUN npm link sol2uml --only=production
```
Comment (rimrakhimov) [inline Dockerfile]: We can group those statements and run them together: `RUN apt install bzip2 && npm install phantom && npm link sol2uml --only=production`

Potentially, it helps to minimize number of layers, however not sure how helpful it may be (https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#minimize-the-number-of-layers)

File: src/main.rs [hunk 10]
```diff
@@ -1,3 +1,14 @@
-fn main() {
-    println!("Hello, world!");
+use sol_to_uml::{config::Config, run};
+
+#[derive(PartialEq, Default, Clone, Debug)]
+struct Commit {
+    hash: String,
+    message: String,
+}
```
Comment (rimrakhimov) [inline src/main.rs]: What is the reason of that structure?

File: src/config.rs [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+#[derive(Deserialize, Clone)]
+#[serde(default)]
+pub struct UMLCreatorConfiguration {
+    pub tmp_dir: PathBuf,
+}
+
+impl Default for UMLCreatorConfiguration {
+    fn default() -> Self {
+        Self {
+            tmp_dir: PathBuf::from("./tmp"),
```
Comment (rimrakhimov) [inline src/config.rs]: If you wish to use temp directory, I suggest to use the one OS provides. You can use `std::env::temp_dir()` for that

File: src/config.rs [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    fn default() -> Self {
+        Self {
+            addr: ("0.0.0.0:8043").parse().expect("should be valid url"),
+        }
+    }
+}
+
+#[derive(Deserialize, Clone)]
+#[serde(default)]
+pub struct UMLCreatorConfiguration {
```
Comment (rimrakhimov) [inline src/config.rs]:
```suggestion
pub struct UmlCreatorConfiguration {
```
https://rust-lang.github.io/api-guidelines/naming.html#casing-conforms-to-rfc-430-c-case

File: example_config.toml [hunk 13]
```diff
@@ -0,0 +1,4 @@
+[server]
+addr = "0.0.0.0:8043"
+[uml_creator]
+tmp_dir = "./sol_to_uml/tmp"
```
Comment (rimrakhimov) [inline example_config.toml]: Why do we actually need to configure that value? Would we somehow may reuse it between instances? If not, I suggest just to use a temporary folder specified by OS

File: src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,92 @@
+mod cli;
+pub mod config;
```
Comment (rimrakhimov) [inline src/lib.rs]: I suggest to make the module private and just reexport `pub use config::Config` structure

File: src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+    data: Json<SolToUmlRequest>,
+) -> Result<Json<SolToUmlResponse>, Error> {
+    let data = data.into_inner();
+
+    let unique_name = Uuid::new_v4().to_string();
+    let contract_dir = tmp_dir.get_ref().join(&unique_name);
+    let contract_dir =
+        web::block(move || std::fs::create_dir(contract_dir.clone()).map(|_| contract_dir))
+            .await
+            .map_err(error::ErrorInternalServerError)??;
```
Comment (rimrakhimov) [inline src/lib.rs]: Try to use `tokio::fs::create_dir` instead

Comment (rimrakhimov) [reply src/lib.rs]:
```suggestion
    let contract_dir =
        web::block(move || std::fs::create_dir(contract_dir.clone()).map(|_| contract_dir))
            .await??;
```
Corresponding `actix::ResponseError` implements from trait for `io::Error`, so you may skip `map_err`

File: src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+    for (name, content) in data.sources.iter() {
```
Comment (rimrakhimov) [inline src/lib.rs]:
```suggestion
    for (name, content) in data.sources {
```

It may iterate automatically (but do it by value in that case)

File: src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+        let file_path = contract_dir.join(name.as_path());
```
Comment (rimrakhimov) [inline src/lib.rs]:
```suggestion
        let file_path = contract_dir.join(name);
```
Not needed

File: src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+        let prefix = file_path.parent();
+        if let Some(prefix) = prefix {
+            let prefix = prefix.to_path_buf();
+            web::block(move || std::fs::create_dir_all(prefix))
+                .await
+                .map_err(error::ErrorBadRequest)??;
+        }
+
+        let mut f = web::block(move || std::fs::File::create(file_path))
```
Comment (rimrakhimov) [inline src/lib.rs]: Use `tokio::fs`

File: src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+            .map_err(error::ErrorBadRequest)??;
+    }
+
+    let uml_path = contract_dir.join(format!("{unique_name}.svg"));
+    let status = Command::new("sol2uml")
+        .arg(&contract_dir)
+        .arg("-o")
+        .arg(&uml_path)
+        .status()
+        .expect("sol2uml command failed to start");
```
Comment (rimrakhimov) [inline src/lib.rs]: Here it make sense to use `web::block`, as that may require some time to complete and may block the whole thread.

Also you may leave current variables into the scope via:
```
let status = {
    let contract_dir = contract_dir.clone();
    let uml_path = uml_path.clone();
    ...
}
```

**Review by rimrakhimov — COMMENTED** [review]
Have a look at https://docs.rs/tempfile/latest/tempfile/struct.TempDir.html#resource-leaking. I believe it may be a great way to create temporary folders in your case

File: Dockerfile (hunk 9 above)
Comment (DanglSan) [reply Dockerfile]: Thx, I`ve tested this and it reduced no-cache docker-compose build time by approx. 1 minute.

File: src/main.rs (hunk 10 above)
Comment (DanglSan) [reply src/main.rs]: Can't believe I missed this, this structure is left over from some tests I did while making this project, so it's completely useless now.

File: src/lib.rs (hunk 19 above)
Comment (DanglSan) [reply src/lib.rs]: Yes, fixed it

File: example_config.toml (hunk 13 above)
Comment (DanglSan) [reply example_config.toml]: I checked the link you gave about [TempDir](https://docs.rs/tempfile/latest/tempfile/struct.TempDir.html#resource-leaking) and rewrote all code related to using temp directories. So now we are using generated temp directory in `std::env::temp_dir()` and I`ve removed this param from config.

File: src/lib.rs (hunk 15 above)
Comment (DanglSan) [reply src/lib.rs]: Ok, I replaced all `web::block` with `tokio::fs`, also used `tokio::process` instead of `std::process` to run console command.

**Comment by rimrakhimov** [comment]
@DanglSan Thank you for changes. 

Regarding tests, in general I suggest to create at least on integration test, which will run the server, send some contracts to be visualized and check the output. However, if I get it right, running such a test would require for the client to have `sol2uml` installed which is probably not a case for most developers. For now I suggest to check manually that everything works as expected and ignore automatic tests. If we decide still to use `sol2uml` after your research about `sol2uml` internal implementation, we would think how tests would be better to be implemented.

Imo, we can merge this PR as it is

**Review by rimrakhimov — APPROVED** [review]

File: Dockerfile [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+# sol2uml needed phantom which installation needed bzip2
+RUN apt-get update && apt-get install bzip2 \
+    && npm install phantom \
+    && npm link sol2uml --only=production
+
+# Copy from the previous build
+COPY --from=build /sol_to_uml/target/release/sol_to_uml /usr/src/sol_to_uml
+
+# Run the binary
+CMD ["/usr/src/sol_to_uml"]
```
Comment (leviska) [inline Dockerfile]: if you intend to run this binary, it's better to make this an `ENTRYPOINT [...]`, not `CMD`

File: docker-compose.yml [hunk 21]
```diff
@@ -0,0 +1,8 @@
+version: '3.3'
+services:
+  server:
+    ports:
+      - "8043:8043"
+    build:
```
Comment (leviska) [inline docker-compose.yml]: `build: .` (and remove context and dockerfile) ?

File: src/lib.rs [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+use tempfile::TempDir;
+use tokio::{io::AsyncWriteExt, process::Command};
+use types::{SolToUmlRequest, SolToUmlResponse};
+
+async fn sol_to_uml(data: Json<SolToUmlRequest>) -> Result<Json<SolToUmlResponse>, Error> {
+    let data = data.into_inner();
+    let contract_dir = TempDir::new()?;
+    let contract_path = contract_dir.path();
+
+    for (name, content) in data.sources {
```
Comment (leviska) [inline src/lib.rs]: you can parallel this loop using `tokio::spawn`
something like
```rust
let join = data.sources.into_iter().map(|(name, content)| {
 tokio::spawn(async move {
   // cycle body
 })
});
futures::future::join_all(join);

```

File: src/lib.rs [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+        Ok(Json(SolToUmlResponse { uml_diagram }))
+    } else {
+        Err(error::ErrorBadRequest(""))
+    }
+}
+
+pub async fn run(config: Config) -> std::io::Result<()> {
+    let socket_addr = config.server.addr;
+
+    log::info!("Sol-to-uml server is starting at {}", socket_addr);
```
Comment (leviska) [inline src/lib.rs]: we've decided to use lowercase letters at the start of the log

File: src/lib.rs (hunk 22 above)
Comment (leviska) [reply src/lib.rs]: actually you can make it even better
instead of `spawn`, use `spawn_blocking` (it still will return some kind of join handle), and then change all the `tokio::fs` operations to `std::fs`
because `tokio::fs` is basically calling `spawn_blocking` in each function, and it should be slower, than calling it once (per file)
so it should look something like
```rust
let join = data.sources.into_iter().map(|(name, content)| {
 tokio::task::spawn_blocking(async move {
    let file_path = contract_path.join(name);
    let prefix = file_path.parent();
    if let Some(prefix) = prefix {
        std::fs::create_dir_all(prefix)?;
    }
    let mut f = std::fs::File::create(file_path)?;
    f.write_all(content.as_bytes())?;
 })
});
futures::future::join_all(join);
```

File: src/lib.rs [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+            let prefix = file_path.parent();
+            if let Some(prefix) = prefix {
+                std::fs::create_dir_all(prefix)?;
+            }
+            let mut f = std::fs::File::create(file_path)?;
+            f.write_all(content.as_bytes())?;
+            Ok(())
+        })
+    });
+    let results: Vec<Result<Result<_, StdError>, JoinError>> =
```
Comment (leviska) [inline src/lib.rs]: I think you can just write `Vec<_>`?

**Review by leviska — APPROVED** [review]
